File: memo/src/main/java/com/example/memo/MemoAdapter.java
package com.example.memo;

import android.app.AlertDialog;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class MemoAdapter extends RecyclerView.Adapter<MemoAdapter.MemoViewHolder> {

    private List<Memo> memoList;
    private Context context;
    private static final int MAX_CONTENT_LENGTH = 50;
    private boolean isMultiSelectMode = false; // 是否处于多选模式
    private List<Memo> selectedItems = new ArrayList<>(); // 保存被选中的备忘录

    public MemoAdapter(List<Memo> memoList, Context context) {
        this.memoList = memoList;
        this.context = context;
    }

    @NonNull
    @Override
    public MemoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.memo_item, parent, false);
        return new MemoViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MemoViewHolder holder, int position) {
        Memo memo = memoList.get(position);
        holder.titleTextView.setText(memo.getTitle());

        // 限制内容显示前50字符，并设置最大行数为2行
        String content = memo.getContent();
        String displayContent = content.length() > MAX_CONTENT_LENGTH
                ? content.substring(0, MAX_CONTENT_LENGTH) + "..."
                : content;
        holder.contentTextView.setText(displayContent);
        holder.contentTextView.setMaxLines(2);

        // 如果是多选模式，显示 CheckBox；否则隐藏
        holder.checkBox.setVisibility(isMultiSelectMode ? View.VISIBLE : View.GONE);
        holder.checkBox.setChecked(selectedItems.contains(memo));

        // 设置内容点击事件，切换内容展开/收起
        holder.contentTextView.setOnClickListener(new View.OnClickListener() {
            private boolean isExpanded = false;

            @Override
            public void onClick(View v) {
                isExpanded = !isExpanded;
                holder.contentTextView.setText(isExpanded ? content : displayContent);
            }
        });

        // 长按启用多选模式
        holder.itemView.setOnLongClickListener(v -> {
            isMultiSelectMode = true;
            notifyDataSetChanged();
            return true;
        });

        holder.itemView.setOnLongClickListener(v -> {
            isMultiSelectMode = true;
            if (context instanceof MainActivityMemo) {
                ((MainActivityMemo) context).toggleDeleteButton(true); // 显示删除按钮
            }
            notifyDataSetChanged();
            return true;
        });


        // 处理多选状态切换
        holder.checkBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                selectedItems.add(memo);
            } else {
                selectedItems.remove(memo);
            }
        });
    }

    @Override
    public int getItemCount() {
        return memoList.size();
    }

    // 批量删除选中 Memo
    public void deleteSelectedMemos() {
        memoList.removeAll(selectedItems);
        selectedItems.clear();
        isMultiSelectMode = false;
        notifyDataSetChanged();
        if (context instanceof MainActivityMemo) {
            ((MainActivityMemo) context).toggleDeleteButton(false); // 隐藏删除按钮
        }
    }


    public static class MemoViewHolder extends RecyclerView.ViewHolder {
        TextView titleTextView;
        TextView contentTextView;
        CheckBox checkBox;

        public MemoViewHolder(@NonNull View itemView) {
            super(itemView);
            titleTextView = itemView.findViewById(R.id.memo_title);
            contentTextView = itemView.findViewById(R.id.memo_content);
            checkBox = itemView.findViewById(R.id.checkbox); // 添加一个 CheckBox
        }
    }
}

File: memo/src/main/java/com/example/memo/MemoAdapter.kt
package com.example.memo

import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.os.Build
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.TextView
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.RecyclerView

class MemoAdapter(
    private val memoList: MutableList<Memo>,
    private val context: Context,
    private val memoDAO: MemoDAO
) : RecyclerView.Adapter<MemoAdapter.MemoViewHolder>() {

    companion object {
        private const val MAX_CONTENT_LENGTH = 50
    }

    var isMultiSelectMode = false
        private set

    private val selectedItems = mutableListOf<Memo>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MemoViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.memo_item, parent, false)
        return MemoViewHolder(view)
    }

    override fun onBindViewHolder(holder: MemoViewHolder, position: Int) {
        val memo = memoList[position]
        holder.titleTextView.text = memo.title
        holder.updateTimeTextView.text = "上次更新: ${memo.updateTime}"

        holder.imageTextIndicator.visibility = 
            if (memo.imagePaths.isNotEmpty()) View.VISIBLE else View.GONE

        val content = memo.content
        val displayContent = if (content.length > MAX_CONTENT_LENGTH) {
            content.substring(0, MAX_CONTENT_LENGTH) + "..."
        } else {
            content
        }
        holder.contentTextView.text = displayContent
        holder.contentTextView.maxLines = 2

        try {
            when (memo.fontName) {
                "宋体" -> ResourcesCompat.getFont(context, R.font.simsunch)
                "宋体（繁）" -> ResourcesCompat.getFont(context, R.font.simsunhk)
                "宋体（港繁）" -> ResourcesCompat.getFont(context, R.font.simsunpro)
                "仿宋" -> ResourcesCompat.getFont(context, R.font.fasimsunch)
                "仿宋（繁）" -> ResourcesCompat.getFont(context, R.font.fasimsunhk)
                "仿宋（港繁）" -> ResourcesCompat.getFont(context, R.font.fasimsunpro)
                "黑体" -> ResourcesCompat.getFont(context, R.font.simheich)
                "黑体（繁）" -> ResourcesCompat.getFont(context, R.font.simheihk)
                "黑体（港繁）" -> ResourcesCompat.getFont(context, R.font.simheipro)
                else -> Typeface.DEFAULT
            }?.let { typeface ->
                holder.contentTextView.typeface = typeface
            }
        } catch (e: Exception) {
            e.printStackTrace()
            holder.contentTextView.typeface = Typeface.DEFAULT
        }

        holder.checkBox.visibility = if (isMultiSelectMode) View.VISIBLE else View.GONE
        holder.checkBox.isChecked = selectedItems.contains(memo)

        holder.itemView.setOnClickListener {
            val intent = Intent(context, MemoDetailActivity::class.java).apply {
                putExtra("memo_id", memo.id)
                putExtra("memo_title", memo.title)
                putExtra("memo_content", memo.content)
                putExtra("memo_update_time", "${memo.updateTime} 修改")
                putStringArrayListExtra("memo_image_paths", ArrayList(memo.imagePaths))
                putExtra("memo_font_name", memo.fontName)
            }
            context.startActivity(intent)
        }

        holder.contentTextView.setOnClickListener(object : View.OnClickListener {
            private var isExpanded = false

            override fun onClick(v: View?) {
                isExpanded = !isExpanded
                holder.contentTextView.text = if (isExpanded) content else displayContent
            }
        })

        holder.itemView.setOnLongClickListener {
            if (!isMultiSelectMode) {
                isMultiSelectMode = true
                if (context is MainActivityMemo) {
                    context.toggleDeleteButton(true)
                }
                notifyDataSetChanged()
            }
            true
        }

        holder.checkBox.setOnClickListener {
            if (holder.checkBox.isChecked) {
                selectedItems.add(memo)
            } else {
                selectedItems.remove(memo)
            }
        }
    }

    override fun getItemCount(): Int = memoList.size

    fun deleteSelectedMemos() {
        for (memo in selectedItems) {
            memoDAO.deleteMemo(memo.id)
        }
        memoList.removeAll(selectedItems)
        selectedItems.clear()
        isMultiSelectMode = false
        notifyDataSetChanged()

        if (context is MainActivityMemo) {
            context.toggleDeleteButton(false)
        }
    }

    fun exitMultiSelectMode() {
        isMultiSelectMode = false
        selectedItems.clear()
        notifyDataSetChanged()
        if (context is MainActivityMemo) {
            context.toggleDeleteButton(false)
        }
    }

    class MemoViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val titleTextView: TextView = itemView.findViewById(R.id.memo_title)
        val contentTextView: TextView = itemView.findViewById(R.id.memo_content)
        val checkBox: CheckBox = itemView.findViewById(R.id.checkbox)
        val updateTimeTextView: TextView = itemView.findViewById(R.id.memo_update_time)
        val imageTextIndicator: TextView = itemView.findViewById(R.id.image_text_indicator)
    }
}
